File: vol2/ch19/LeetCode46-2.cpp
class Solution {
    vector<vector<int>> ans;        //存放答案
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n=nums.size();
        vector<int> x=nums;
        dfs(x,0);
        return ans;
    }
    void dfs(vector<int>&x,int i) {     //回溯算法
        int n=x.size();
        if (i>=n)
         ans.push_back(x);
        else {
            for(int j=i;j<n;j++) {
                swap(x[i],x[j]);        //交换x[i]与x[j]
                dfs(x,i+1);
                swap(x[i],x[j]);        //回溯：交换x[i]与x[j]
            }
        }
    }
};

File: vol2/ch19/LeetCode46-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.ans,x=[],nums
        self.dfs(x,0)
        return self.ans
    def dfs(self,x,i):  #回溯算法
        n=len(x)
        if i>=n:
            self.ans.append(list(x))    #等同于x深拷贝
        else:
            for j in range(i,n):
                x[i],x[j]=x[j],x[i]     #交换x[i]与x[j]
                self.dfs(x,i+1)
                x[i],x[j]=x[j],x[i]     #回溯：交换x[i]与x[j]
